docs/proposals/failover/application-failover.md

```diff
@@ -0,0 +1,562 @@
+avoid invalid failovers due to misconfigurations. In such cases, we need a parameter to init the failover behavior.
+Optional configuration items include:
+
+* Healthy State. We just want to migrate those applications that have run successfully. For those applications that failed at the beginning, we prefer to solve them in the single cluster.
+* A Period of time. We can wait a period of time before the application failover is initiated.
+
+For some short-lived exceptions(which can be recovered after a period of time), we don't want to immediately migrate the application to other clusters which may make the migration very frequent.
+Therefore, we need a parameter to tolerate the unhealthy application.
+
+### API change
+
+#### PropagationPolicy API change
+
+```golang
+// PropagationSpec represents the desired behavior of PropagationPolicy.
+type PropagationSpec struct {
+	// ....
+
+	PropagateDeps bool `json:"propagateDeps,omitempty"`
+
+	// Failover indicates how Karmada migrates applications in case of failures.
+	// If this value is nil, failover is disabled.
+	// +optional
+	Failover *FailoverBehavior `json:"failover,omitempty"`
+
+	// ...
+}
+
+// FailoverBehavior indicates failover behaviors in case of an application or
+// cluster failure.
+type FailoverBehavior struct {
+    // Application indicates failover behaviors in case of application failure.
+    // If this value is nil, failover is disabled.
+    // If set, the PropagateDeps should be true so that the dependencies could
+    // be migrated along with the application.
+    // +optional
+    Application *ApplicationFailoverBehavior `json:"application,omitempty"`
+  
+    // Cluster indicates failover behaviors in case of cluster failure.
+    // If this value is nil, failover is disabled.
+    // +optional
+    Cluster *ClusterFailoverBehavior `json:"cluster,omitempty"`
+}
+
+// ApplicationFailoverBehavior indicates application failover behaviors.
+type ApplicationFailoverBehavior struct {
+    // InitiateGates indicates the preconditions of the failover process.
+    // If specified, all gates will be evaluated for prerequisites.
+    // +optional
+    InitiateGates *FailoverInitiateGate `json:"initiateGates,omitempty"`
+  
+    // ActiveGates indicates the conditions of performing the failover process.
+    //
+    // +kubebuilder:validation:MinItems=1
+    // +required
+    ActiveGates *FailoverActiveGate `json:"activeGates"`
+  
+    // PurgeMode represents how to deal with the legacy applications on the
+    // cluster from which the application is migrated.
+    // Valid options are "Immediately", "Graciously" and "Never".
+    // Defaults to "Graciously".
+    //
+    // +kubebuilder:validation:Enum=Immediately;Graciously;Never
+    // +optional
+    PurgeMode string `json:"purgeMode,omitempty"`
+  
+    // EscapeSeconds represents the period of time the cluster from which the
+    // application was migrated from should be escaped.
+    // Defaults to 600s.
```

Comment:
Is the default value too long? Can it be considered to be the same as the default value with failover?